store/branch/store.go

```diff
@@ -0,0 +1,107 @@
+package branch
+
+import (
+	"io"
+
+	"golang.org/x/exp/maps"
+	"slices"
+
+	"cosmossdk.io/store/v2"
+)
+
+var (
+	_ store.KVStore         = (*Store)(nil)
+	_ store.BranchedKVStore = (*Store)(nil)
+)
+
+// Store implements both a KVStore and BranchedKVStore interfaces. It is used to
+// accumulate writes that can be later committed to backing SS and SC engines or
+// discarded altogether. If a read is not found through an uncommitted write, it
+// will be delegated to the SS backend.
+type Store struct {
+	// storage reflects backing storage for reads that are not found in uncommitted volatile state
+	//
+	// XXX/TODO: We use a SC backend here instead of SS since not all SS backends
+	// may support reverse iteration (which is needed for state machine logic).
+	storage store.Tree
+
+	// storeKey reflects the store key used for the store
+	storeKey string
```

Comment:
i thought last week we talked about not making storage store key aware? or is this a different part?

Reply:
We did talk about it yes. The conclusion we came to is: 

* Store key (as a string) is the cleanest way to support the ability for devs to have multiple SC trees.
* Our default implementation doesn't use them -- i.e. they are ignored. However, we need to support the ability for app devs to implement a `RootStore` where there are multiple SC trees, so @fdymylja and I agreed this was the cleanest way.

Reply:
makes sense thank you. should updating the adr be part of this pr?

Reply:
Yup! This will be apart of ADR updates